core/ir/opnd_api.h

```diff
@@ -2271,6 +2289,28 @@ opnd_t
 opnd_create_base_disp_aarch64(reg_id_t base_reg, reg_id_t index_reg,
                               dr_extend_type_t extend_type, bool scaled, int disp,
                               dr_opnd_flags_t flags, opnd_size_t size);
+
+DR_API
+/**
+ * Same as opnd_create_base_disp_shift_aarch64 but creates an operand that uses vector
+ * registers for the base and/or index.
+ * At least one of \p base_reg and \p index_reg should be a vector register.
+ * \p element_size indicates the element size for any vector registers used and must be
```

Comment:
To make sure I understand: this PR is not talking about element sizes of loaded or stored regions of memory (that would seem to want a discussion of generalizing into other memory block breakdowns like LDP or x86 PUSHA): this is only talking about the ADR opcode which constructs a whole bunch of addresses from elements in vector registers, right?  Or is there a plan to use this for scatter/gather too?

If this is only for ADR: could we avoid this IR change altogether by *not* considering ADR with vector registers to have a memory operand at all?  If we treat it as just an arithmetic vector operation than we can use the existing register operand element size.  I know x86 LEA today has a memory operand, partly because that's how it's decoded and encoded: but that causes headaches and we have to have exclusions every time we do client operations on memory operands.

Reply:
> this PR is not talking about element sizes of loaded or stored regions of memory

Correct. Its the size of the elements in the vector register.

> Or is there a plan to use this for scatter/gather too?

Yes. That is the plan. SVE has scatter/gather instructions that use Z registers for the base and index registers. For example:
```
LD1B   { <Zt>.S }, <Pg>/Z, [<Zn>.S{, #<imm>}]
LDNT1B { <Zt>.S }, <Pg>/Z, [<Zn>.S{, <Xm>}]
ST1B   { <Zt>.D }, <Pg>,   [<Xn|SP>, <Zm>.D, <mod>]
```
We plan to use the same mechanism to encode the element size for any memory operands that use vector registers.

Reply:
If it will be used for scatter/gather: the question is whether we could use the same implementation for x86 scatter/gather.  Just 1 bit is added here as there are only 2 sizes.  After talking to @abhinav92003 it does look like that would work for today's x86 scatter/gather which also only have 2 sizes.

Reply:
My understanding is that the memory signature, i.e. vector base + scaled vector index, is unique to the `ADR` instruction:
```
[<Zn>.D,<Zm>.D{,LSL #<sh>}]  <-- 64 bit form

[<Zn>.S,<Zm>.S{,LSL #<sh>}]      \__ 32 bit form
[<Zn>.D,<Zm>.D,(S|U)XTW {#<sh>}] /
```
and is not directly supported by SVE's load/store and prefetch instructions. I could be wrong though, @jackgallagher-arm  has been looking at this more recently than me and my knowledge may be out-of-date.

> this PR is not talking about element sizes of loaded or stored regions of memory (that would seem to want a discussion of generalizing into other memory block breakdowns like LDP or x86 PUSHA): this is only talking about the ADR opcode which constructs a whole bunch of addresses from elements in vector registers, right?

Yes, it's not about _actual memory load/stores_, only the _construction of addresses_ for use in later actual memory load/stores.

> If this is only for ADR: could we avoid this IR change altogether by not considering ADR with vector registers to have a memory operand at all?

We thought having a memory operand with zero bytes transferred would be a convenience to the user, indicating the purpose of the operands in the IR as memory addresses. Like the prefetch instructions which indicate a memory address but don't transfer anything. But it looks like in this case that's not what you would like. If so, what should the IR look like? Just plain registers and scaling modifiers? e.g.
```
adr z0.d, [z0.d, z0.d, SXTW] : adr %z0.d,%z0.d,sxtw -> %z0.d
adr z0.d, [z0.d, z0.d, UXTW] : adr %z0.d,%z0.d,uxtw -> %z0.d

adr z0.s, [z0.s, z0.s] : adr %z0.s,%z0.s -> %z0.s
adr z0.s, [z0.s, z0.s, LSL #1] : adr %z0.s,%z0.s,lsl #1 -> %z0.s
adr z31.s, [z31.s, z31.s, LSL #3] : adr %z31.s,%z31.s,lsl #3 -> %z31.s
```

Reply:
Thinking about it more: since the existing ADR (with GPR's) and ADRP are modeled as memory operands, and LEA on x86 is as well, to be consistent we would have to make a compatibility change and go change all of those as well if we wanted to not have these compute-address opcodes have memory operands.  That may be more than we want to do.  So maybe the solution here is the best way forward.

Reply:
Not sure whether @abhinav92003 or @khuey have different opinions.

Reply:
> So maybe the solution in this PR is the best way forward.

We are happy with this if you are.

> Not sure whether @abhinav92003 or @khuey have different opinions.

Let us know what you think @abhinav92003 @khuey.

Reply:
@abhinav92003 has been traveling/out of office.  Let's just go ahead with this PR as-is.